src/crimson/common/gated.h

```diff
@@ -0,0 +1,67 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <seastar/core/gate.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/future-util.hh>
+
+#include "crimson/common/exception.h"
+#include "crimson/common/log.h"
+#include "include/ceph_assert.h"
+
+namespace crimson::common {
+
+class Gated {
+ public:
+  virtual void print(std::ostream&) const = 0;
+ protected:
+  static seastar::logger& gated_logger() {
+    return crimson::get_logger(ceph_subsys_osd);
+  }
+  template <typename Func>
+  void gated_dispatch(const char* what, Func&& func) {
+    (void) seastar::with_gate(pending_dispatch, std::forward<Func>(func)
+    ).handle_exception([this, what] (std::exception_ptr&& eptr) {
+      if (*eptr.__cxa_exception_type() == typeid(system_shutdown_exception)) {
+	gated_logger().debug(
+	    "{}, {} skipped, system shutdown", *this, what);
+	return seastar::now();
+      }
+      gated_logger().error(
+          "{} gated_dispatch() {} caught exception: {}", *this, what, eptr);
+      ceph_abort("unexpected exception from gated_dispatch()");
+    });
+  }
+  template <typename Func>
+  seastar::future<> gated_dispatch_with_future(const char* what, Func&& func) {
+    return seastar::with_gate(pending_dispatch, std::forward<Func>(func)
+    ).handle_exception([this, what] (std::exception_ptr eptr) {
+      if (*eptr.__cxa_exception_type() == typeid(system_shutdown_exception)) {
+	gated_logger().debug(
+	    "{}, {} skipped, system shutdown", *this, what);
+	return seastar::now();
+      }
+      gated_logger().error(
+          "{} gated_dispatch() {} caught exception: {}", *this, what, eptr);
+      ceph_abort("unexpected exception from gated_dispatch()");
+    });
+  }
+
+  seastar::future<> close_dispatch_gate() {
```

Comment:
```suggestion
  seastar::future<> close() {
```

or just use `stop()`.